### PR TITLE
Ldid

### PR DESCRIPTION
**Pull request by alvarofe**
should fix https://github.com/radare/radare2/issues/4368. I use `rabin2 -OC binary` i can change it to `rabin2 -O ldid binary`  as you said @radare let me know.

### CODE REVIEW

**Comment by radare** [comment]
Travis is unhappy

**Comment by alvarofe** [comment]
updated and fixed some crashes with fuzzed binaries. ldid is totally fucked up when dealing with those files

**Comment by alvarofe** [comment]
btw don't close the issue until i add some test to the repo

**Comment by radare** [comment]
👍

File: libr/bin/format/mach0/mach0.c [hunk 1]
```diff
@@ -381,6 +381,45 @@ static int parse_dysymtab(struct MACH0_(obj_t)* bin, ut64 off) {
...
+	len = r_buf_fread_at (bin->b, off, (ut8*)&link, bin->big_endian ? "4I" : "4i", 1);
+	if (len < 1) {
+		eprintf ("Failed to get data while parsing LC_CODE_SIGNATURE command\n");
+		return;
+	}
+	data =  link.dataoff;
+	if (data > bin->size || data + sizeof(struct super_blob_t) > bin->size)
+	    	return;
+	struct super_blob_t *super = (struct super_blob_t *) (bin->b->buf + data);
+	count = r_swap_ut32(super->count);
```
Comment (zamaudio) [inline libr/bin/format/mach0/mach0.c]: Why do you cast some memory address to a struct and then swap endian?  Smells bad to me.

**Comment by alvarofe** [comment]
@zamaudio i changed to your new API. Regarding the other what is wrong with casting to that struct?

**Comment by zamaudio** [comment]
First of all, if your data is a stream of bytes, you should always read it as a sequence of linear bytes, don't swap them.  Then, if you need to interpret some of those bytes as integers, you need to know which endian YOU want to store them in, and call r_read_bleXX() depending on the target endianness you want.  If you are trying to read integer values bigger than 1 byte by CASTING a stream to a struct and allowing the machine to choose which endian to interpret them in, you're doing it wrong.  You need to parse the values one by one and store them in a known endian to the struct.

**Comment by alvarofe** [comment]
Basically what i am doing with the casting is that in that addr there is a struct of type `blob_t` and then i read the value using the api you recently created to read from big endian.

```c
struct blob_t *entitlements = (struct blob_t *) ((ut8*)super + begin);
len = r_read_ble32 (&entitlements->length, true) - sizeof(struct blob_t);
```
I use directly true because even when we have bin->big_endian (value we get when parsing the file) in the file codesign.c apple uses `ntohl` so it's independent from the file

```c
uint32_t offset = ntohl(sb->index[count].offset);
```
https://opensource.apple.com/source/Security/Security-55471/sec/Security/Tool/codesign.c

But it's true i never thought about these issues that is why i have to check the mach parse to use bin->big_endian along your api to fix some issues in the future with mach files with different MAGIC value

**Comment by radare** [comment]
you can also load the struct using the fmt apis, a bit slower but safer.


(r_buf_fread_at)

**Comment by alvarofe** [comment]
let me change that and update the PR

**Comment by alvarofe** [comment]
the thing is that the struct has inside an array of other structs

```
struct super_blob_t {
    struct blob_t blob;
    ut32 count;
    struct blob_index_t index[];
}; 
```
for me let it as is and wait to @revskills :P

File: binr/rabin2/rabin2.c [hunk 2]
```diff
@@ -358,7 +358,9 @@ static int rabin_do_operation(const char *op) {
 		RBinPlugin *plg = r_bin_file_cur_plugin (cur);
 		if (!plg) break;
 		if (plg->signature) {
-			plg->signature (cur);
+			char *sign = plg->signature (cur);
```
Comment (radare) [inline binr/rabin2/rabin2.c]: maybe const char * or memleak?

File: libr/core/bin.c [hunk 3]
```diff
@@ -2109,7 +2109,8 @@ static int bin_signature(RCore *r, int mode) {
 	RBinPlugin *plg = r_bin_file_cur_plugin (cur);
 	if (!plg) return false;
 	if (plg->signature) {
-	    	plg->signature (cur);
+	    	char *signature = plg->signature (cur);
+		r_cons_printf ("%s\n", signature);
```
Comment (radare) [inline libr/core/bin.c]: memleak?

File: binr/rabin2/rabin2.c (hunk 2 above)
Comment (alvarofe) [inline binr/rabin2/rabin2.c]: there shouldn't be memleak as long is just return bin->signature and when freeing the mach bin object it does the free. I can add const char * + plus casting. 1m

**Comment by alvarofe** [comment]
updated @radare

**Comment by trufae** [comment]
this is not documented in rabin2 -O help

**Comment by alvarofe** [comment]
done

**Pull request merged**